src/rgw/rgw_admin.cc

```diff
@@ -5504,6 +5509,126 @@ int main(int argc, const char **argv)
+    RGWObjectCtx obj_ctx(store);
+
+    //Get bucket info
+    RGWBucketInfo bucket_info;
+    RGWSysObjectCtx sys_ctx = store->svc.sysobj->init_obj_ctx();
+    map<string, bufferlist> attrs;
+
+    //split tenant/bucket
+    auto pos = bucket_name.find('/');
+    if (pos != boost::string_ref::npos) {
+      tenant = bucket_name.substr(0, pos);
+      bucket_name = bucket_name.substr(pos + 1);
+    }
+
+    //Get bucket info
+    RGWBucketInfo bucket_info;
+    RGWSysObjectCtx sys_ctx = store->svc.sysobj->init_obj_ctx();
+    map<string, bufferlist> attrs;
+    ret = store->get_bucket_info(sys_ctx, tenant, bucket_name, bucket_info, NULL, &attrs);
+    if (ret < 0) {
+      cerr << " Bucket info failed: tenant: "<< tenant << "bucket_name: " << bucket_name << " " << ret << std::endl;
+      return ret;
+    }
+
+    //Get User info - for display name
+    RGWUserInfo user_info;
+    ret = rgw_get_user_info_by_uid(store, bucket_info.owner, user_info);
+    if (ret < 0) {
+      cerr << " User info failed: "<< ret << std::endl;
+      return ret;
+    }
+
+    RGWRados::Bucket target(store, bucket_info);
+    RGWRados::Bucket::List list_op(&target);
+    int max = 1000;
+
+    list_op.params.list_versions = true;
+    list_op.params.allow_unordered = true;
+    list_op.params.marker = marker;
+
+    bool is_truncated = false;
+
+    do {
+      objs.clear();
+      ret = list_op.list_objects(max, &objs, &common_prefixes, &is_truncated);
+      if (ret < 0) {
+        cerr << " List objects failed: "<< ret << std::endl;
+        return ret;
+      }
+
+      //Loop through the results
+      for (const auto& obj : objs) {
+
+        const rgw_obj r_obj(bucket_info.bucket, obj.key);
+        ret = get_obj_attrs(store, obj_ctx, bucket_info, r_obj, attrs);
+        if (ret < 0){
+          cerr << "Get object attrs failed" << ret << std::endl;
+          return (ret);
+        }
+        const auto& aiter = attrs.find(RGW_ATTR_ACL);
+        if (aiter == attrs.end()){
+          cerr << "No acls found for object. Continuing with the next object" << std::endl;
+          continue;
+        } else {
+          bufferlist& bl = aiter->second;
+          RGWAccessControlPolicy policy(g_ceph_context);
+          ACLOwner owner;
+          try {
+            decode(policy, bl);
+            owner = policy.get_owner();
+          } catch (buffer::error& err) {
+            cerr << " Decode policy failed: " << std::endl;
+            return -EIO;
+          }
+
+          //Get the ACL from the policy
+          RGWAccessControlList& acl = policy.get_acl();
+
+          //Remove grant that is set to old owner
+          acl.remove_canon_user_grant(owner.get_id());
+
+          //Create a grant and add grant
+          ACLGrant grant;
+          grant.set_canon(bucket_info.owner, user_info.display_name, RGW_PERM_FULL_CONTROL);
+          acl.add_grant(&grant);
+
+          //Add the ACL back to policy
+          policy.set_acl(acl);
```

Comment:
since `RGWAccessControlList& acl` is a non-const reference to `RGWAccessControlPolicy::acl`, you're already modifying it in place - this extra `set_acl()` function shouldn't be necessary